### ISSUE TITLE
STRtree: include temporal data in index

### ISSUE THREAD

**Issue by adamjstewart** [Toblerity/Shapely#1165]
STRtree is very useful for querying a set of geometries to find polygons that overlap the query polygon. However, many polygons also contain temporal information.

In my use case, I have a directory of satellite imagery files. I would like to create an STRtree with the bounding box for all images, then use the index to find images that correspond to a certain geospatial location and range in time. Is this possible with the current STRtree implementation, or would it need to be somehow extended to support this? Or would that need to happen at the GEOS level?

One alternative would be to use STRtree for the geospatial indexing and keep a separate BST just for time, then index both and compute the intersection. But I'm not sure how easy this would be to do.

@calebrob6

**Comment by mwtoews** [comment]
One suggestion is to look into [rtree](https://github.com/Toblerity/rtree), which supports [3D indexes](https://rtree.readthedocs.io/en/latest/tutorial.html#d-indexes). You can store start/end times (e.g.) in days since 1970-01-01 into the third dimension. `intersection` should do the right thing, but `nearest` might have unexpected behavior, as it will attempt to evaluate a distance with mixed space/time units. Also, I think there is no way to make a 2D query with at 3D index if you wanted to ignore time, so you'd need to make a second index if that were wanted.

**Comment by adamjstewart** [comment]
Yes, I'm currently using rtree with 3D indexes. However, rtree doesn't support things like Polygons or coordinate systems. If I just want to use the bounding box I can project everything to the right coordinate system, but if I want to use arbitrary polygons things become problematic.
